### DISCUSSION TITLE
AudioStreamPlayer should use internal channels for automatic polyphony

### DISCUSSION THREAD

**Discussion by elvisish** [godotengine/godot-proposals#5705] [Engine Core]
AudioStreamPlayers can only play one sound at once, which means the user has to set-up multiple AudioStreamPlayers for overlapping sounds, or rapid-fire sounds that shouldn't cut each other off. If they used internal channels to play the sounds on, the user could define a polyphony for each AudioStreamPlayer and it could automatically find a free channel to play the sound on, without interrupting the current sound. The channel *could* be manually defined, but by default it could just be chosen and if all of the channels are currently playing, the AudioStreamPlayer could just go back to the earliest channel and stop the stream for use.

It could also be easier to allow for automatically crossfade between channels, having the current channel fade out while the new channel fades in, but all still on the same AudioStreamPlayer.

**Comment by Calinou** [comment]
Automatic polyphony support is already implemented in 4.0 :slightly_smiling_face: 

It won't be backported to 3.x though, as it relies on a rewrite of the audio system.

**Reply by elvisish** [reply]
Really, within the same AudioPlayer?

**Reply by Calinou** [reply]
Yes :slightly_smiling_face:

**Reply by elvisish** [reply]
> It won't be backported to 3.x though, as it relies on a rewrite of the audio system.

Yeah I noticed in one of the proposals it was such a major change it couldn't be backported. I'm currently juggling sounds between multiple players in an array and using tweens to fade some sounds, I assume this is the only way it can be handled in 3.x?

And I still think a lot of this is relevant as 4.0 isn't in stable yet and there will be many people using 3.x for a while to come.

**Reply by Calinou** [reply]
>  I'm currently juggling sounds between multiple players in an array and using tweens to fade some sounds, I assume this is the only way it can be handled in 3.x?

You can create AudioStreamPlayer nodes as needed, then `queue_free()` them once the sound is done playing. An autoload like [this one](https://github.com/Calinou/escape-space/blob/master/autoload/sound.gd) can do it for you.

**Reply by elvisish** [reply]
> >  I'm currently juggling sounds between multiple players in an array and using tweens to fade some sounds, I assume this is the only way it can be handled in 3.x?
> 
> 
> 
> You can create AudioStreamPlayer nodes as needed, then `queue_free()` them once the sound is done playing. An autoload like [this one](https://github.com/Calinou/escape-space/blob/master/autoload/sound.gd) can do it for you.

Is that more expensive than cycling pre-instanced nodes?

**Reply by Calinou** [reply]
> Is that more expensive than cycling pre-instanced nodes?

Yes, but only marginally so. It's very unlikely it'll make a noticeable difference, even on mobile. You'd need to be playing thousands of sounds at once, which will already overwhelm the audio mixer anyway.
